fix: File watcher doesn't notify the deleted event of old path after changing case of filename

hyy215 opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/eclipse-theia/theia/blob/master/CONTRIBUTING.md
-->

#### What it does
File watcher notify the deleted event of old path and the added event of current path after renaming filename.

But `fsp.realpath(oldPath)` return the current path after changing case of filename, so file watcher can't return deleted event of old path.

Skip `fsp.realpath(oldaPath)` when the path is deleted, and return the real path of old path.

#### How to test
1. Watch files change
```ts
@injectable()
export class FileWatcherTest {
    @inject(FileService)
    protected readonly fileService: FileService;

    init() {
        this.fileService.onDidFilesChange(event => {
            console.log(event);
        }
    }    
}
```
2. Rename test.txt to TEST.txt
3. notify the deleted event of old path and the added event of current path, like this:
```
[
  {
    uri: "file:///f%3A/Demo/test.txt",
    type: 2,
  },
  {
    uri: "file:///f%3A/Demo/TEST.txt",
    type: 1,
  },
]
```

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/eclipse-theia/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to review in accordance with [the review guidelines](https://github.com/eclipse-theia/theia/blob/master/doc/pull-requests.md#reviewing)

vince-fugnitto reviewed (commented):
Thank you for your contribution :+1: 

In order to accept your changes please be sure to sure to sign the [eclipse contributor agreement (eca)](https://github.com/eclipse-theia/theia/blob/master/CONTRIBUTING.md#eclipse-contributor-agreement) with the same email as your authorship. Please be sure to use your real name and email.

hyy215 commented:
> Thank you for your contribution 👍
> 
> In order to accept your changes please be sure to sure to sign the [eclipse contributor agreement (eca)](https://github.com/eclipse-theia/theia/blob/master/CONTRIBUTING.md#eclipse-contributor-agreement) with the same email as your authorship. Please be sure to use your real name and email.

Modified real name and email is my real email.

vince-fugnitto commented:
> @vince-fugnitto The email is my real email and I changed to real name.

No problem, please let us know when you have completed the signing of the ECA and we will re-validate it.
At the moment it is still failing:

![image](https://user-images.githubusercontent.com/40359487/172164963-3c15a8bc-13c4-4a1a-988c-2cf691ef6768.png)


```diff
From 6991c4d72b7ab3ede1340307ee091f235da8bda3 Mon Sep 17 00:00:00 2001
From: hyy <hyy215215@163.com>
Date: Thu, 2 Jun 2022 17:05:52 +0800
Subject: [PATCH] fix: File watcher doesn't notify the deleted event of old
 path after changing case of filename

Signed-off-by: huangyanyan <hyy215215@163.com>
```

vince-fugnitto reviewed (commented):
I'm a little bit confused as to what the changes are attempting to fix, maybe @paul-marechal you have an idea?

From what I can tell I am correctly getting the `deleted` event when verifying master:

```ts
this.fs.onDidFilesChange(event => {
    console.dir(event);

    // Deleted.
    event.getDeleted().forEach(d => {
        console.log(`deleted: ${d.resource.toString()}`);
    });
});
```

When a file is renamed from `a.ts` to `A.ts` I see the log:

```
2022-06-07T13:31:06.032Z root INFO deleted: file:///home/workspace/test/a.ts
```

I also see (simplified version):

```
[
    {
        uri: "file:///home/workspaces/test/A.ts",
        type: 0 // FileChangeType.UPDATED
    },
    {
        uri: "file:///home/workspaces/test/a.ts",
        type: 2 // FileChangeType.DELETED
    }
]
```

hyy215 commented:
> I'm a little bit confused as to what the changes are attempting to fix, maybe @paul-marechal you have an idea?
> 
> From what I can tell I am correctly getting the `deleted` event when verifying master:
> 
> ```ts
> this.fs.onDidFilesChange(event => {
>     console.dir(event);
> 
>     // Deleted.
>     event.getDeleted().forEach(d => {
>         console.log(`deleted: ${d.resource.toString()}`);
>     });
> });
> ```
> 
> When a file is renamed from `a.ts` to `A.ts` I see the log:
> 
> ```
> 2022-06-07T13:31:06.032Z root INFO deleted: file:///home/workspace/test/a.ts
> ```
> 
> I also see (simplified version):
> 
> ```
> [
>     {
>         uri: "file:///home/workspaces/test/A.ts",
>         type: 0 // FileChangeType.UPDATED
>     },
>     {
>         uri: "file:///home/workspaces/test/a.ts",
>         type: 2 // FileChangeType.DELETED
>     }
> ]
> ```

@vince-fugnitto maybe this is related to [fsPromises.realpath](https://nodejs.org/dist/latest-v16.x/docs/api/fs.html#fspromisesrealpathpath-options)，my operating system is win10.
When a file is renamed from a.ts to A.ts, nsfw fire event，like this
```
[
  {
    action: 1,
    directory: "E:\\ldp-project\\test1\\service",
    file: "a.ts",
  },
  {
    action: 3,
    directory: "E:\\ldp-project\\test1\\service",
    oldFile: "a.ts",
    newDirectory: "E:\\ldp-project\\test1\\service",
    newFile: "A.ts",
  },
  {
    action: 2,
    directory: "E:\\ldp-project\\test1\\service",
    file: "A.ts",
  },
]
```
https://github.com/eclipse-theia/theia/blob/b2e44a9367b1d9c9b1a0977edf57deab3cc50759/packages/filesystem/src/node/nsfw-watcher/nsfw-filesystem-service.ts#L245-L303
then `handleNsfwEvents` will resolve event path and transform nsfw.ActionType to FileChangeType，but the result of `fsp.realpath(a.ts)`  is `A.ts`， so resolve result like it(simplified)
```
[{
    uri: 'A.ts',
    type: 2,  // FileChangeType.DELETED
}, {
    uri: 'A.ts',
    type: 2,  // FileChangeType.DELETED
}, {
    uri: 'A.ts',
    type: 1,  // FileChangeType.ADDED
}, {
    uri: 'A.ts',
    type: 0,  // FileChangeType.UPDATED
}]
```
https://github.com/eclipse-theia/theia/blob/b2e44a9367b1d9c9b1a0977edf57deab3cc50759/packages/filesystem/src/node/file-change-collection.ts#L33-L78
then normalize same file changes
```
// 1th normalize result
uri: A.ts
type: 2,  // FileChangeType.DELETED

// 2th normalize result
uri: A.ts
type: 2,  // FileChangeType.DELETED

//  3th normalize result
uri: A.ts
type: 0,  // FileChangeType.UPDATED

// 4th normalize result
uri: A.ts
type: 0,  // FileChangeType.UPDATED
```
finally, only return A.ts updated event，no a.ts deleted event

vince-fugnitto commented:
> my operating system is win10.

@hyy215 unfortunately I don’t have a windows machine to confirm the fix, I’ll let another contributor continue with the review 👍

msujew reviewed (approved):
It's a bit hard to test this, since there's some weird issue (related to https://github.com/eclipse-theia/theia/issues/11308) related to the file watcher not being correctly propagated/updated. But anyway, as far as I can tell, this seems to address the issue at hand. When renaming a file, I receive these notifications:

```
[
  {
    "path": "C:\\Users\\Mark\\Source\\Repos\\langium\\test.txt",
    "type": "DELETED"
  },
  {
    "path": "C:\\Users\\Mark\\Source\\Repos\\langium\\TEST.txt",
    "type": "ADDED"
  }
]
```

For comparison, this is what I'm getting on `master`:

```
[
  {
    "path": "C:\\Users\\Mark\\Source\\Repos\\langium\\TEST.txt",
    "type": "UPDATED"
  }
]
```

paul-marechal commented:
@hyy215 thanks for your contribution by the way!

We just merged an alternative change, which I believe is better as it simplifies our code.
